File: content/LLD/Reference-Code/17. Decorater Pattern/bad_example.py
from abc import ABC, abstractmethod


class Beverage(ABC):
    @abstractmethod
    def get_description(self) -> str:
        pass

    @abstractmethod
    def get_cost(self) -> int:
        pass


class Coffee(Beverage):
    def get_description(self):
        return "Plain coffee"

    def get_cost(self):
        return 20


class CoffeeWithMilk(Coffee):
    def get_description(self):
        return "Plain coffee with Milk"

    def get_cost(self):
        return 30


coffee1 = Coffee()
print(coffee1.get_description())
print(coffee1.get_cost())

coffee1 = CoffeeWithMilk()
print(coffee1.get_description())
print(coffee1.get_cost())

File: content/LLD/Reference-Code/17. Decorater Pattern/BadExample.java
public class BadExample {
    interface Beverage {
        String getDescription();
        int getCost();
    }

    static class Coffee implements Beverage {
        @Override
        public String getDescription() {
            return "Plain coffee";
        }

        @Override
        public int getCost() {
            return 20;
        }
    }

    // Class explosion: creating a subclass for every combination
    static class CoffeeWithMilk implements Beverage {
        @Override
        public String getDescription() {
            return "Plain coffee with Milk";
        }

        @Override
        public int getCost() {
            return 30;
        }
    }

    public static void main(String[] args) {
        Coffee coffee1 = new Coffee();
        System.out.println(coffee1.getDescription());
        System.out.println(coffee1.getCost());

        CoffeeWithMilk coffee2 = new CoffeeWithMilk();
        System.out.println(coffee2.getDescription());
        System.out.println(coffee2.getCost());
    }
}
